Add acpi device resource enumeration support

najumon1980 opened the pull request:
Add acpi device resource enumeration interface and updated uart driver with the new acpi interface

najumon1980 committed:
arch: x86: add interface for encode irq flags

add interface for encode irq flags from acpica to arch specfic.
Currently enabled only for x86 archiecture.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
lib: acpi: add device resource enum support

add device resource enumaration support such as irq and mmio.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

dcpleung commented on drivers/serial/uart_ns16550.c:
```diff
@@ -1243,6 +1318,14 @@ static const struct uart_driver_api uart_ns16550_driver_api = {
+	}
+
+#define UART_NS16550_IRQ_CONFIG_PCIE0(n)                                                           \
+	_CONCAT(UART_NS16550_IRQ_CONFIG_ACPI, DT_INST_ON_BUS(n, acpi))(n)
```
What if there are no ACPI instances but purely PCIe instances?

najumon1980 replied:
The design change takecare UART instances on both ACPI and PCI or DTS and ACPI at the same time. (For example: UART0 on ACPI node, UART1 on PCI or UART0 on DTS and UART1 on PCI. But doesn't support ACPI and DTS at the same time.

dcpleung commented on drivers/serial/uart_ns16550.c:
```diff
@@ -1243,6 +1318,14 @@ static const struct uart_driver_api uart_ns16550_driver_api = {
+#define UART_NS16550_IRQ_CONFIG_ACPI1(n)                                                           \
+	static void irq_config_func##n(const struct device *dev)                                   \
+	{                                                                                          \
+	}
```
Why is this empty for ACPI1?

najumon1980 replied:
For instances on ACPI node, both IRQ and MMIO config as part of init routine. So ACPI irq config is add as stub function.

najumon1980 replied:
updated the code

dcpleung reviewed (changes requested):
I haven't looked in details. But took a quick look over the UART driver, and it seems you are replacing PCIe with ACPI in several places. That is not right.

najumon1980 commented:
> I haven't looked in details. But took a quick look over the UART driver, and it seems you are replacing PCIe with ACPI in several places. That is not right.

The design change takecare UART instances on both ACPI and PCI or DTS and ACPI at the same time. (For example: UART0 on ACPI node, UART1 on PCI or UART0 on DTS and UART1 on PCI. But doesn't support ACPI and DTS at the same time. You can go through existing code and changes to get full picture.

dcpleung commented:
This is getting overly complicated. Could you find another way of doing this? Might want to use `COND_CODE_1()` instead of macro trampolines.

najumon1980 commented:
> This is getting overly complicated. Could you find another way of doing this? Might want to use `COND_CODE_1()` instead of macro trampolines.

I total agree that it will complicate drivers if we need to support on multiple bus topologies and for this I just created a draft PR (WIP): https://github.com/zephyrproject-rtos/zephyr/pull/62993 to abstract such complexity and is given current just for analyzes purpose.

najumon1980 committed:
lib: acpi: add resource enumeration shell command

add resource enumeration and acpi method shell commands such as
retrieve mmio and interrupt resources.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

tbursztyka commented on dts/bindings/acpi/acpi-vbus.yaml:
```diff
@@ -0,0 +1,14 @@
+
+include: base.yaml
+
+bus: acpi
```
No ACPI is _not_ a hw bus. This is not compliant with DTS.

najumon1980 replied:
Agree that ACPI is not a hardware bus, but it defined to standardizes OS interface with platform HW and FW which is not limited to power management but include planform configuration such as interrupt routing, DMA, device MMIO resources, device enumeration and many more. ACPI driver itself have registers sets and ISR to mange its run time functionalities.  So in implementation point of view, it is more like a driver which deals with HW and its associated interface. For more details of ACPI: https://uefi.org/htmlspecs/ACPI_Spec_6_4_html/01_Introduction/Introduction.html
 
I believe  the question which is debatable here is DTS vs ACPI, that is whether we can completely move to ACPI base model for drivers used in platforms like x86 which include device enumeration apart from resource enumerations. But currently Zephyr moving to DTS base including application interface and most of the test apps started using dts base interface such as DEVICE_DT_GET instead of device_get_binding().  So removing DTS support for some platform is not viable currently without complicate these test apps.

Following option we may need to think:
- Option 1:  Continue the DTS base device enumeration (discover devices in a platform) for ACPI devices just like currently how PCI device enumerated and will be consistent with existing Zephyr driver's application/usage model, but all other MMIO/IRQ resource enumeration will be via ACPI like PCI devices. This need some bus node to distinguish the driver's parent node in DTS like pcie. Currently I follow this option and provide a virtual bus node name "acpi".

- Option 2: Move to a complete ACPI base model for all non PCI devices in x86 platforms which include device enumerations and without any DTS support. Here PCI driver interface also need to move completely on PCI based device enumeration (Device discovery)

najumon1980 replied:
removed ACPI node. In order to distinguish  whether a driver use ACPI interface or not, one can use DT_NODE_HAS_PROP(n, acpi_uid)). Please refer PR(https://github.com/zephyrproject-rtos/zephyr/pull/62993) for more details of the usage.

@galak could you please have look on this PR as well as PR(https://github.com/zephyrproject-rtos/zephyr/pull/62993) ?

najumon1980 replied:
@tbursztyka @dcpleung could you please review the PR ?

najumon1980 committed:
drivers: pci: update prt retrieve based on pnp id

update prt retrieve based on acpi pnp id instead of acpi device
path/name

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
boards: x86: add pci controller node with acpi pnp id

add acpi pnp/hw id for pcie node to enable support for retreive
interrupt routing information for pci legacy interrupt via acpi

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

jhedberg reviewed (changes requested):
qemu_x86_64 doesn't boot (well, at least any shell sample doesn't boot) with this PR if you enable PCIE and ACPI. We pretty badly need an upstream CI test app for ACPI that runs on qemu. Anyway, please make sure that you provide the new PNP IDs for all platforms that need them, and also remove the now seemingly unnecessary PCI root device Kconfig option.

tbursztyka commented:
DNM label added for now

tbursztyka commented on include/zephyr/acpi/acpi.h:
```diff
@@ -39,6 +57,44 @@ struct acpi_mcfg {
 	uint64_t _reserved;
 	struct acpi_mcfg_allocation pci_segs[];
 } __packed;
+struct acpi_irq_resource {
```
add an empty line before

tbursztyka commented on include/zephyr/acpi/acpi.h:
```diff
@@ -39,6 +57,44 @@ struct acpi_mcfg {
 } __packed;
+struct acpi_irq_resource {
+	uint32_t flags;
+	uint8_t irq_vector_max;
```
move this attribute at the end of the structure

tbursztyka commented on include/zephyr/acpi/acpi.h:
```diff
@@ -39,6 +57,44 @@ struct acpi_mcfg {
+};
+
+struct acpi_mmio_resource {
+	uint8_t mmio_max;
```
same here

tbursztyka commented on lib/acpi/acpi.c:
```diff
@@ -501,6 +505,106 @@ ACPI_RESOURCE *acpi_resource_parse(ACPI_RESOURCE *res, int res_type)
+		if (res->Data.Irq.InterruptCount > CONFIG_ACPI_IRQ_VECTOR_MAX) {
+			return -ENOMEM;
+		}
+		irq_res->irq_vector_max = res->Data.Irq.InterruptCount;
```
add an empty line before

tbursztyka commented on lib/acpi/acpi.c:
```diff
@@ -501,6 +505,106 @@ ACPI_RESOURCE *acpi_resource_parse(ACPI_RESOURCE *res, int res_type)
+
+		res = ACPI_NEXT_RESOURCE(res);
+		if (mmio_cnt >= CONFIG_ACPI_MMIO_ENTRIES_MAX &&
+				res->Type != ACPI_RESOURCE_TYPE_END_TAG) {
```
indentation is wrong, it should be below last '(' + 1 space

tbursztyka commented on lib/acpi/Kconfig:
```diff
@@ -58,3 +58,15 @@ config ACPI_HID_LEN_MAX
 	help
 	  Size of HID string.
+
+config ACPI_IRQ_VECTOR_MAX
```
shouldn't we be able to use kernel dynamic allocation, so these 2 config would not be necessary?

tbursztyka commented on lib/acpi/acpi_shell.c:
```diff
@@ -220,18 +228,101 @@ static int dump_prt(const struct shell *sh, size_t argc, char **argv)
+
+	for (int i = 0; i < 1000; i++) {
+		dev = acpi_device_by_index_get(i);
+
```
you can remove that empty line

tbursztyka commented on lib/acpi/acpi_shell.c:
```diff
@@ -259,6 +350,42 @@ static int read_table(const struct shell *sh, size_t argc, char **argv)
+		shell_error(sh, "ACPI get table failed: %d\n", status);
+		return status;
+	}
+	if (acpi_obj.Type == ACPI_TYPE_INTEGER) {
```
add an empty line before

tbursztyka commented on lib/acpi/acpi_shell.c:
```diff
@@ -259,6 +350,42 @@ static int read_table(const struct shell *sh, size_t argc, char **argv)
+	} else {
+		shell_print(sh, "Return info type: %d\n", acpi_obj.Type);
+	}
+	return 0;
```
add an empty line before

tbursztyka commented on include/zephyr/acpi/acpi.h:
```diff
@@ -86,15 +86,25 @@ struct acpi_mmio_resource {
  * @return The HID of the ACPI node
  */
-#define ACPI_DT_HID(node_id) DT_PROP_OR(node_id, hardware_id, "INVALD")
+#define ACPI_DT_HID(node_id) DT_PROP_OR(node_id, acpi_hw_id, "INVALD")
```
"INVALD" ?  (typo?)

tbursztyka commented on lib/acpi/acpi.c:
```diff
@@ -373,6 +360,19 @@ static int acpi_early_init(void)
+	}
+	/* TODO: assume platform have only one PCH with single PCI bus (bus 0). */
+	return acpi_get_irq_table(&bus_ctx, child_dev->path,
+		bus_ctx.pci_prt_table, sizeof(bus_ctx.pci_prt_table));
```
indentation is wrong

tbursztyka commented on drivers/pcie/host/pcie.c:
```diff
@@ -4,6 +4,13 @@
+#if defined(CONFIG_X86_64)
+#define DT_DRV_COMPAT intel_pcie
+#else
+#error "ACPI DSDT not supported by current arch"
```
so we could get rid of that message (which is a bit confusing anyway)

najumon1980 replied:
updated

tbursztyka commented on drivers/pcie/host/pcie.c:
```diff
@@ -4,6 +4,13 @@
  */
+#if defined(CONFIG_PCIE_PRT)
+#if defined(CONFIG_X86_64)
+#define DT_DRV_COMPAT intel_pcie
```
why not DT_DRV_COMPAT pcie_controller  ? (and thus, getting rid of the CONFIG_X86_64 above).

Up to the relevant architecture to set CONFIG_PCIE_PRT accordingly.

najumon1980 replied:
updated

jhedberg replied:
> why not DT_DRV_COMPAT pcie_controller ? (and thus, getting rid of the CONFIG_X86_64 above).
> 
> Up to the relevant architecture to set CONFIG_PCIE_PRT accordingly.

@tbursztyka maybe this is just me not understanding how DTS works, but we don't have any pcie nodes using a "pcie,controller" compat string. We do have a `pice-controller.yaml` that PCIe bindings include, but does that mean that our tooling generates DT macros to match that file as well?

tbursztyka replied:
see my answer in the comment below

tbursztyka commented on drivers/pcie/host/pcie.c:
```diff
@@ -533,6 +540,12 @@ static int pcie_init(void)
 	};
 
+#ifdef CONFIG_PCIE_PRT
+	if (acpi_legacy_irq_init(
```
not sure it's the right place to call this actually, we could just initialize it all in acpi code.

najumon1980 replied:
We need PCI Bus ACPI HW/PNP ID from DTS for retrieving during init. Later call is only per device which just return cached value.

najumon1980 commented:
> qemu_x86_64 doesn't boot (well, at least any shell sample doesn't boot) with this PR if you enable PCIE and ACPI. We pretty badly need an upstream CI test app for ACPI that runs on qemu. Anyway, please make sure that you provide the new PNP IDs for all platforms that need them, and also remove the now seemingly unnecessary PCI root device Kconfig option.

I updated board/dts to support older platforms (including QEMU). But I don't have board currently for testing APL and EHL (QEMU already tested). If you have board then please verify and let me know if possible.

najumon1980 commented on include/zephyr/acpi/acpi.h:
```diff
@@ -39,6 +57,54 @@ struct acpi_mcfg {
+ * @param node_id DTS node identifier
+ * @return The HID of the ACPI node
+ */
+#define ACPI_DT_UID(node_id) DT_PROP_OR(node_id, acpi_uid, "INVALD")
```
I already added debug ASSERT (run time for such failer) in pcie.c and are you still prefer it use DT_PROP instead of DT_PROP_OR ?

finikorg replied:
I think @jhedberg proposal is better since the app would not build if there is no attribute, no assert needed.

jhedberg commented:
> I updated board/dts to support older platforms (including QEMU). But I don't have board currently for testing APL and EHL (QEMU already tested). If you have board then please verify and let me know if possible.

I can test on APL and EHL. However, I just tested `qemu_x86_64` and it doesn't work, at least the shell seems to get stuck if you enable ACPI, PCIE & PRT support in Kconfig. I used this command line:
```
west build -b qemu_x86_64 samples/subsys/shell/shell_module -t run -- -DCONFIG_SYS_HEAP_RUNTIME_STATS=y -DCONFIG_ACPI=y -DCONFIG_PCIE=y -DCONFIG_DYNAMIC_INTERRUPTS=y -DCONFIG_PCIE_PRT=y
```

jhedberg commented:
@najumon1980 I tested on APL (up_squared), and unfortunately `samples/subsys/shell/shell_module` doesn't work (with default configuration - I didn't add anything extra like with my earlier qemu command line). The app just gets stuck here:
```
Jumping to Entry Point: 0x1153 (48 31 c0 48 31 d2 48)
```
However, with current upstream main branch it works fine:
```
Jumping to Entry Point: 0x1153 (48 31 c0 48 31 d2 48)


uart:~$
```

finikorg commented on lib/acpi/Kconfig:
```diff
@@ -14,12 +14,6 @@ module = ACPI
 module-str = acpi
 source "subsys/logging/Kconfig.template.log_config"
 
-config ACPI_PRT_BUS_NAME
```
You need to remove this config from all boards configurations

najumon1980 replied:
updated

finikorg commented on boards/x86/qemu_x86/qemu_x86.dts:
```diff
@@ -50,6 +50,8 @@
 		#address-cells = <1>;
 		#size-cells = <1>;
 		compatible = "intel,pcie";
+		acpi-hw-id = "PNP0A08";
```
Should we include dt header and use `PCI_EXPRESS_ROOT_HID_STRING` or similar?

najumon1980 replied:
it may not same for future platforms

jhedberg replied:
>> Should we include dt header and use `PCI_EXPRESS_ROOT_HID_STRING` or similar?
> it may not same for future platforms

Isn't this a standard value though, i.e. I don't think it will change:
https://github.com/acpica/acpica/blob/fee7180f7c1e4cc986440c34a6df559a479421ce/source/common/ahids.c#L283

jhedberg replied:
> I suggest it is better to have a common dtsi file say "intel_x86_common.dtsi" and include all nodes which are common for all Intel platforms and then include it in platform specific dtsi files.

That's not a bad idea, although we already have at least a little bit of this kind of "avoid duplication" by having SoC-specific dtsi files which then all boards that use the same SoC include in their dts files.

This is a separate issue from @finikorg's proposal though. Even if we had a common dtsi file for x86 it would still make sense to have defines for the standard PNP IDs (at least the commonly used ones).

finikorg commented on lib/acpi/acpi_shell.c:
```diff
@@ -220,18 +228,100 @@ static int dump_prt(const struct shell *sh, size_t argc, char **argv)
 
-	if (argc < 2) {
+	if (argc < 3) {
+		shell_error(sh, "Invalid Arugements [Eg: acpi enum PNP0103 0]\n");
```
Arugements -> Arguments

jhedberg commented:
@najumon1980 now that `tests/lib/acpi/` is merged, please updated it in the same go to test the PRT stuff using the updated ACPI API. Feel free to also add new tests there, if you want.

jhedberg commented on drivers/pcie/host/pcie.c:
```diff
@@ -533,6 +535,12 @@ static int pcie_init(void)
 
+#ifdef CONFIG_PCIE_PRT
+	if (acpi_legacy_irq_init(
+		ACPI_DT_HID(DT_DRV_INST(0)), ACPI_DT_UID(DT_DRV_INST(0))) != 0) {
```
@finikorg seems like this is related to https://github.com/zephyrproject-rtos/zephyr/pull/62694#discussion_r1379158394

I.e. was the DT_DRV_COMPAT changed to the wrong value?

najumon1980 replied:
yes, we need to give "DT_DRV_COMPAT intel_pcie"

najumon1980 commented:
> > I updated board/dts to support older platforms (including QEMU). But I don't have board currently for testing APL and EHL (QEMU already tested). If you have board then please verify and let me know if possible.
> 
> I can test on APL and EHL. However, I just tested `qemu_x86_64` and it doesn't work, at least the shell seems to get stuck if you enable ACPI, PCIE & PRT support in Kconfig. I used this command line:
> 
> ```
> west build -b qemu_x86_64 samples/subsys/shell/shell_module -t run -- -DCONFIG_SYS_HEAP_RUNTIME_STATS=y -DCONFIG_ACPI=y -DCONFIG_PCIE=y -DCONFIG_DYNAMIC_INTERRUPTS=y -DCONFIG_PCIE_PRT=y
> ```

You may need to apply following PRs:
https://github.com/acpica/acpica/pull/905
https://github.com/zephyrproject-rtos/zephyr/pull/64570

jhedberg commented:
> You may need to apply following PRs: [acpica/acpica#905](https://github.com/acpica/acpica/pull/905) #64570

Why is there no change to `west.yml` in this PR if it depends on an update to a module? You should follow the process documented here:
https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules

Or when you say "may" do you mean that you're not sure if it's required or not?

I'd suggest to direct a PR to the `zephyr` branch of our acpica mirror - we can then even merge it if everyone agrees on it. The key thing is that you need to demonstrate that CI passes on this PR, and that includes qemu_x86_64 also working (like it does right now with the upstream tree).

jhedberg commented:
@najumon1980 I did some googling, and found this regarding qemu ACPI & PCI details. It's written from a Linux perspective, but should be helpful to us too:
https://gist.github.com/mcastelino/96e5fb1e47ffd5ec5b99bb5843855769

jhedberg commented:
Another very useful document I found:
https://docs.kernel.org/PCI/acpi-info.html

It e.g. states "PCI host bridges are PNP0A03 or PNP0A08 devices."

From the above, it sounds like we may not have to encode this stuff in DTS, rather can have fixed defines in the code itself. This is what both FreeBSD and Linux seem to do too.

najumon1980 commented:
> > You may need to apply following PRs: [acpica/acpica#905](https://github.com/acpica/acpica/pull/905) #64570
> 
> Why is there no change to `west.yml` in this PR if it depends on an update to a module? You should follow the process documented here: https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules
> 
> Or when you say "may" do you mean that you're not sure if it's required or not?
> 
> I'd suggest to direct a PR to the `zephyr` branch of our acpica mirror - we can then even merge it if everyone agrees on it. The key thing is that you need to demonstrate that CI passes on this PR, and that includes qemu_x86_64 also working (like it does right now with the upstream tree).

Looks you missed the 2nD PR my comment. I give two PRs name (https://github.com/acpica/acpica/pull/905 and 
https://github.com/zephyrproject-rtos/zephyr/pull/64570)

For your comment on "may": it is because the issue will not able to repro if platform use TSC timer (instead of HPET) and hence it may not require for ADL etc. But QEMU use HPET and hence those fixes required.

najumon1980 commented:
> Another very useful document I found: https://docs.kernel.org/PCI/acpi-info.html
> 
> It e.g. states "PCI host bridges are PNP0A03 or PNP0A08 devices."
> 
> From the above, it sounds like we may not have to encode this stuff in DTS, rather can have fixed defines in the code itself. This is what both FreeBSD and Linux seem to do too.

With ACPI and PCI we can get rid of DTS altogether, but in Zephyr we still need DTS because of existing drivers so tightly coupled with it. Anyway, we use DTS then I believe it is the best place to keeping ACPI ID instead of hard coding in the driver code.

jhedberg commented:
> Looks you missed the 2nd PR on my comment. I give two PRs name ([acpica/acpica#905](https://github.com/acpica/acpica/pull/905) and #64570)
> 
> For your comment on "may": it is because the issue will not able to repro if platform use TSC timer (instead of HPET) and hence it may not require for ADL etc. But QEMU use HPET and hence those fixes required.

Thanks for the clarification! So we really need to get the timer issue sorted out and merged before coming back to this PR. One thing that's unclear to me: what part of ACPICA functionality (that you're trying to use in this PR) depends on AcpiOsGetTimer() being fully functional?

finikorg commented:
> For your comment on "may": it is because the issue will not able to repro if platform use TSC timer (instead of HPET) and hence it may not require for ADL etc. But QEMU use HPET and hence those fixes required.

We can also use `+tsc-deadline` in the `-cpu` argument list (but it seems requite to recompile qemu TCG or use `-enable-kvm` and use '-cpu=host`.

jhedberg commented:
> We can also use `+tsc-deadline` in the Qemu `-cpu` argument list (but it seems requite to recompile qemu TCG or use `-enable-kvm` and use `-cpu=host`.

I just tried to add this to the command line (in `board.cmake`), and it worked without any recompilation. I suppose it depends what installation of qemu is used - mine is version 8.1.2 from "brew" on macOS.

finikorg commented:
> > We can also use `+tsc-deadline` in the Qemu `-cpu` argument list (but it seems requite to recompile qemu TCG or use `-enable-kvm` and use `-cpu=host`.
> 
> I just tried to add this to the command line (in `board.cmake`), and it worked without any recompilation. I suppose it depends what installation of qemu is used - mine is version 8.1.2 from "brew" on macOS.

Does hello_world with APIC timer selected work in qemu after you enable this flag?

jhedberg commented:
> Does hello_world with APIC timer selected work in qemu after you enable this flag?

Yes. Even the shell worked.

najumon1980 committed:
arch: x86: add a sperate timer interface for acpica lib

add a separate timer interface for acpica lib instead of using
system timer which might use driver interface such as HPET and this
might cause init priority issue if a driver which need to init
before system timer driver instantiate.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
manifest: update west point to latest acpica commit

update west to latest acpica commit for timer interface changes

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -40,6 +58,55 @@ struct acpi_mcfg {
+ */
+#define ACPI_DT_HID(node_id) DT_PROP_OR(node_id, acpi_hw_id, "INVALID")
+
+/**
+ * @brief Get the ACPI HID for a node
+ *
+ * @param node_id DTS node identifier
+ * @return The HID of the ACPI node
+ */
+#define ACPI_DT_UID(node_id) DT_PROP_OR(node_id, acpi_uid, "INVALID")
```
Looks like the documentation is wrong (probably copy-paste mistake from `ACPI_DT_HID()`. Also as previously noted, please don't use `DT_PROP_OR()` rather deal with he lack of the property in the code itself using `DT_NODE_HAS_PROP()` or similar (possibly combined with an assert)

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -434,6 +419,20 @@ uint32_t acpi_legacy_irq_get(pcie_bdf_t bdf)
+	struct acpi_dev *child_dev = acpi_device_get(hid, inst);
+
+	if (!child_dev) {
+		LOG_ERR("no such PCI bus device %s %s\n", hid, inst);
```
unnecessary `\n` at the end

jhedberg commented on lib/acpi/acpi_shell.c:
```diff
@@ -104,7 +104,14 @@ static void dump_dev_res(const struct shell *sh, ACPI_RESOURCE *res_lst)
 			shell_print(sh, "ACPI_RESOURCE_TYPE_EXTENDED_ADDRESS64");
 			break;
 		case ACPI_RESOURCE_TYPE_EXTENDED_IRQ:
+			ACPI_RESOURCE_EXTENDED_IRQ * ext_irq_res = &res->Data.ExtendedIrq;
```
Incorrect style: there should be no space between `*` and the variable name. There's a false-positive Compliance error because of this, which you should solve the same way as some other already existing case statements do, i.e. add a `{` at the end of the case statement.

jhedberg commented on lib/acpi/acpi_shell.c:
```diff
@@ -104,7 +104,14 @@ static void dump_dev_res(const struct shell *sh, ACPI_RESOURCE *res_lst)
+			ACPI_RESOURCE_EXTENDED_IRQ * ext_irq_res = &res->Data.ExtendedIrq;
+
 			shell_print(sh, "ACPI_RESOURCE_TYPE_EXTENDED_IRQ");
+			shell_print(sh, "Triggering:%x, Polarity:%x, Shareable:%x,",
+				    ext_irq_res->Triggering, ext_irq_res->Polarity,
+				    ext_irq_res->Shareable);
+			shell_print(sh, "InterruptCount:%d, Interrupts[0]:%x",
+				    ext_irq_res->InterruptCount, ext_irq_res->Interrupts[0]);
```
Please use the same style as for other resources, i.e. each line after `ACPI_RESOURCE_TYPE_...` is indented with a tab (`\t`). Also, each variable on its own line, with a space after `:`.

jhedberg commented on lib/acpi/acpi_shell.c:
```diff
@@ -220,19 +225,93 @@ static int dump_prt(const struct shell *sh, size_t argc, char **argv)
+			continue;
+		}
+
+		shell_print(sh, "%d) Name:%s, HID:%s", i, dev->path ? dev->path : "Non",
```
Please use the same format as on line 241, i.e. have a space after `:`

jhedberg commented on lib/acpi/acpi_shell.c:
```diff
@@ -220,19 +225,93 @@ static int dump_prt(const struct shell *sh, size_t argc, char **argv)
+	}
+
+	if (dev->path) {
+		shell_print(sh, "Device Path:\n \t%s", dev->path);
```
Does the path really need to be on a new line after "Device Path:"?

jhedberg commented on lib/acpi/acpi_shell.c:
```diff
@@ -220,19 +225,93 @@ static int dump_prt(const struct shell *sh, size_t argc, char **argv)
+
+			shell_print(sh, "Device MMIO resources");
+			for (int i = 0; i < ACPI_RESOURCE_COUNT_GET(&mmio_res); i++) {
+				shell_print(sh, "\tType: %x, Address:%p, Size:%d",
```
Please be consistent with your style: you have a space after "Type:" but not after "Address:" or "Size:".

jhedberg commented on lib/acpi/acpi_shell.c:
```diff
@@ -220,19 +225,93 @@ static int dump_prt(const struct shell *sh, size_t argc, char **argv)
+
+			shell_print(sh, "Device IRQ resources");
+			for (int i = 0; i < irq_res.irq_vector_max; i++) {
+				shell_print(sh, "\tIrq Num: %x, Flags:%x", irq_res.irqs[i],
```
Same here: missing space after "Flags:". Also, I'd capitalize IRQ for consistency (like you do two lines earlier).

jhedberg commented on lib/acpi/acpi_shell.c:
```diff
@@ -189,28 +196,26 @@ static int dump_dev_prs(const struct shell *sh, size_t argc, char **argv)
+			shell_print(sh, "PCI Legacy IRQ number for bus %d device %d function %d is: %d",
+				 bus, dev_num, func, irq);
+		} else {
+			shell_print(sh, "PCI Legacy IRQ number for bus %d device %d function %d Not found",
```
Compliance is complaining about the too long line. You could shorten it a bit by using "dev" and "func" instead of the full words

jhedberg reviewed (changes requested):
Good progress, thanks for doing the rebase! There are some minor things, but some bigger issues too. It also seems you're missing updates to `tests/lib/acpi` (if you don't update the app there's no way the CI checks will pass)

najumon1980 committed:
tests: lib: acpi: update test cases for dev and resource enum

update acpi test cases with device enumeration, mmio and irq
resources enumeration.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -345,11 +348,12 @@ int acpi_current_resource_free(ACPI_RESOURCE *res)
-static int acpi_get_irq_table(char *bus_name, ACPI_PCI_ROUTING_TABLE *rt_table, uint32_t rt_size)
+static int acpi_get_irq_table(char *bus_name,
+				ACPI_PCI_ROUTING_TABLE *rt_table, uint32_t rt_size)
 {
 	ACPI_BUFFER rt_buffer;
 	ACPI_NAMESPACE_NODE *node;
-	ACPI_STATUS status;
+	int status;
```
Why are you changing this? `ACPI_STATUS` should be the correct type.

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -455,6 +456,107 @@ ACPI_RESOURCE *acpi_resource_parse(ACPI_RESOURCE *res, int res_type)
+	struct acpi_reg_base *reg_base = mmio_res->reg_base;
+	int mmio_cnt = 0;
+
+	memset(mmio_res, 0, sizeof(struct acpi_mmio_resource));
```
minor, but it's shorter and more explicit about what it does:
```c
memset(mmio_res, 0, sizeof(*mmio_res));
```

The above will stay correct even if you change the type of `mmio_res`.

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -455,6 +456,107 @@ ACPI_RESOURCE *acpi_resource_parse(ACPI_RESOURCE *res, int res_type)
+
+	do {
+		if (!res->Length) {
+			LOG_DBG("Error: zero length found!\n");
```
Unnecessary `\n`

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -40,6 +58,55 @@ struct acpi_mcfg {
+ * @param node_id DTS node identifier
+ * @return The HID of the ACPI node
+ */
+#define ACPI_DT_HID(node_id) DT_PROP_OR(node_id, acpi_hw_id, "INVALID")
+
+/**
+ * @brief Get the ACPI UID for a node
+ *
+ * @param node_id DTS node identifier
+ * @return The UID of the ACPI node
+ */
+#define ACPI_DT_UID(node_id) DT_PROP_OR(node_id, acpi_uid, "INVALID")
```
Same thing I've mentioned before: don't hide the missing property by using `DT_PROP_OR()`, rather push the handling to the user of these macros with the help of e.g. `DT_INST_NODE_HAS_PROP()` or similar.

jhedberg reviewed (changes requested):
@najumon1980 looks better, and now the test cases also pass on qemu_x86_64! However, still some things to be fixed - please see my inline comments.

jhedberg commented:
@najumon1980 looks like there are some Kconfig leftovers that still need to be removed, e.g:
```
warning: ACPI_PRT_BUS_NAME (defined at boards/x86/intel_rpl/Kconfig.defconfig:38) defined without a type
```

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -360,29 +364,26 @@ static int acpi_get_irq_table(char *bus_name, ACPI_PCI_ROUTING_TABLE *rt_table,
 	}
 
 	rt_buffer.Pointer = rt_table;
-	rt_buffer.Length = rt_size * sizeof(ACPI_PCI_ROUTING_TABLE);
+	rt_buffer.Length = rt_size;
```
Why this change? I provided the justifications for passing `ARRAY_SIZE()` instead of `sizeof()` to this API in the commit message of commit 149b00f70cf8845006799edb86344a12901c53db

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -434,6 +421,20 @@ uint32_t acpi_legacy_irq_get(pcie_bdf_t bdf)
+
+	/* TODO: assume platform have only one PCH with single PCI bus (bus 0). */
+	return acpi_get_irq_table(child_dev->path, acpi.pci_prt_table,
+							  sizeof(acpi.pci_prt_table));
```
This should use `ARRAY_SIZE()` (see my other comment)

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -434,6 +421,20 @@ uint32_t acpi_legacy_irq_get(pcie_bdf_t bdf)
 	return UINT_MAX;
 }
+
+int acpi_legacy_irq_init(char *hid, char *inst)
```
Shouldn't the second parameter be called `uid` instead of `inst`?

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -513,10 +615,11 @@ int acpi_device_type_get(ACPI_RESOURCE *res)
 }
 
-struct acpi_dev *acpi_device_get(char *hid, int inst)
+struct acpi_dev *acpi_device_get(char *hid, char *inst)
```
The second parameter should probably be called `uid`, right?

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -513,10 +615,11 @@ int acpi_device_type_get(ACPI_RESOURCE *res)
 	struct acpi_dev *child_dev;
-	int i = 0, inst_id;
+	int i = 0;
+	char *inst_id;
```
Maybe `uid_str` instead of `inst_id`? Also, since variables should be declared in the smallest possible scope where they're used, you should probably move this declaration inside the branches later in the function.

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -749,6 +852,23 @@ ACPI_MADT_LOCAL_APIC *acpi_local_apic_get(int cpu_num)
+	ACPI_STATUS status;
+	ACPI_BUFFER ret_buff;
+
+	ret_buff.Length = sizeof(ACPI_OBJECT);
```
minor: here also it would be better to use `sizeof(*ret_obj)` since that way there's no chance of using a different type than what `ret_obj` actually is.

jhedberg commented on lib/acpi/acpi_shell.c:
```diff
@@ -220,19 +226,93 @@ static int dump_prt(const struct shell *sh, size_t argc, char **argv)
+{
+	struct acpi_dev *dev;
+
+	for (int i = 0; i < 1000; i++) {
```
The magic `1000` seems a bit arbitrary. Where's that coming from? It should probably at least be some `ACPI_MAX_DEV` define.

jhedberg commented on tests/lib/acpi/src/main.c:
```diff
@@ -17,14 +17,39 @@ ZTEST(acpi, test_mcfg_table)
-	status = acpi_get_irq_routing_table(CONFIG_ACPI_PRT_BUS_NAME,
-					    irq_prt_table, ARRAY_SIZE(irq_prt_table));
-	zassert_ok(status, "Failed to get PRT");
+	dev = acpi_device_get("PNP0A08", "0");
```
These strings shouldn't be hardcoded here. *If* they're provided as part of DTS (like you're right now doing) then you should get them from DTS. If they don't come from DTS, then you should be using defines for them (but that would mean removing this stuff from DTS too).

Since this is not PCIe driver code, you can't use the DT_INST APIs. There are two other options: if we assume all platforms use the `pcie0` node label for their PCIe bus node, then you can use `DT_NODELABEL(pcie0)` instead of `DT_DRV_INST(0)`. However, if we don't want to make this assumption, you could e.g. require DTS overlays for each board in this app, and provide DT aliases in the overlay. After that you could use `DT_ALIAS()` to get the PCIe bus node (if you want to see examples, just do something like `git grep aliases tests/`).

najumon1980 replied:
updated with test case enable with conditional build.

jhedberg commented on tests/lib/acpi/src/main.c:
```diff
@@ -17,14 +17,39 @@ ZTEST(acpi, test_mcfg_table)
+	struct acpi_mmio_resource mmio_res;
+	int ret;
+
+	dev = acpi_device_get("PNP0B00", "0");
```
Same here. In this case I think a DT alias is the right way to go since it's getting more platform specific (seems like this is the HID value for "AT Real-Time Clock" according to ACPICA)

najumon1980 replied:
updated with test case enable with conditional build. This test case builds only if board rtc node have acpi hid property.

jhedberg commented on drivers/pcie/host/pcie.c:
```diff
@@ -291,8 +303,17 @@ unsigned int pcie_alloc_irq(pcie_bdf_t bdf)
 	    irq >= CONFIG_MAX_IRQ_LINES ||
 	    arch_irq_is_used(irq)) {
 
+	/*
+	 * In some platforms, PCI interrupts are hardwired to specific interrupt inputs
```
Minor: I think the consistent code comment style is to begin the text right after `/* ...` instead of the next line.

jhedberg commented on lib/acpi/acpi_shell.c:
```diff
@@ -274,8 +354,13 @@ SHELL_STATIC_SUBCMD_SET_CREATE(
-		  "read acpi table (for read mad table, eg:acpi read_table APIC)",
+	SHELL_CMD(enum_all, NULL, "enumerate all device in acpi name space (eg:acpi enum_all)",
+		  enum_all_dev),
+	SHELL_CMD(
+		dev_res, NULL,
```
Unnecessary line split

jhedberg commented on tests/lib/acpi/src/main.c:
```diff
@@ -17,14 +17,43 @@ ZTEST(acpi, test_mcfg_table)
 }
 
-ZTEST(acpi, test_irq_routing_table)
+#if ACPI_DT_HAS_HID(DT_NODELABEL(pcie0))
+ZTEST(acpi, test_dev_enum)
 {
```
Please don't use this style to skip tests. Instead, use `ztest_test_skip()`, e.g.:
```c
	if (!ACPI_DT_HAS_HID(DT_NODELABEL(pcie0))) {
		test_test_skip();
		return;
	}
```

The reasoning for this is explained here: https://docs.zephyrproject.org/latest/develop/test/ztest.html#skipping-tests

finikorg replied:
There is also Z_TEST_SKIP_IFNDEF(config)

finikorg replied:
defined as  `COND_CODE_1(config, (), (ztest_test_skip()))`

finikorg replied:
@jhedberg ztest_test_skip() should not return

jhedberg replied:
Btw, I can see that there are many places under `tests/` with explicit `return;` statements after `z_test_skip();`, but I guess that's just redundant/dead code then.

finikorg replied:
https://github.com/zephyrproject-rtos/zephyr/blob/a3ff19a39eb18bc10ae38698f5c91ad50c5748d7/subsys/testsuite/ztest/src/ztest.c#L334-L346

finikorg replied:
The test might not compile using this ztest_test_skip() , then `ifdef` should be OK.

jhedberg replied:
> The test might not compile using this ztest_test_skip() , then `ifdef` should be OK.

No, even in that case it's not ok (you can see the reasoning in the Zephyr documentation that I linked). If the code doesn't build, then you should still use `ztest_test_skip()` in a separate pre-processor branch. You could do that by having the pre-processor condition inside the test function body, together with an `#else`, or then by having two variants of the same test function (one of which would simply contain the skip directive). The point with all this is that all functions are always visible in the test statistics & output, and we can see whenever some tests get skipped, instead of them just silently being compiled out without any trace.

jhedberg commented on tests/lib/acpi/src/main.c:
```diff
@@ -17,14 +17,43 @@ ZTEST(acpi, test_mcfg_table)
+}
+#endif
+
+#if ACPI_DT_HAS_HID(DT_NODELABEL(rtc))
+ZTEST(acpi, test_resource_enum)
```
I don't see you adding HID properties to any "rtc" node instances in this pull request, which makes this test essentially dead code. Please include those DTS changes if you want to add such a test case. Also, see my earlier note about using `ztest_test_skip()` for skipping tests on architectures/platforms that cannot support a certain test case.

najumon1980 replied:
> I don't see you adding HID properties to any "rtc" node instances in this pull request, which makes this test essentially dead code. Please include those DTS changes if you want to add such a test case. Also, see my earlier note about using `ztest_test_skip()` for skipping tests on architectures/platforms that cannot support a certain test case.

in order to add rtc node, rtc driver need to update and I have another PR for same (https://github.com/zephyrproject-rtos/zephyr/pull/62993)

jhedberg replied:
> > I don't see you adding HID properties to any "rtc" node instances in this pull request, which makes this test essentially dead code. Please include those DTS changes if you want to add such a test case. Also, see my earlier note about using `ztest_test_skip()` for skipping tests on architectures/platforms that cannot support a certain test case.
> 
> in order to add rtc node, rtc driver need to update and I have another PR for same (#62993)

We should be able to see that the new test case passes in CI (i.e. with qemu), so you need to provide some solution already in this PR. The test case itself is not really related to rtc - all it needs is "some" device with an acpi-hid property. That's why I've been recommending to use a board-specific devicetree overlay for the test together with aliases.

If the board already has some DT node with the acpi-hid property, then you just need aliases. If it doesn't you could either add a new node or enhance an existing one with the property, and then combined that with aliases. So something like:
`tests/lib/acpi/boards/qemu_x86_64.overlay`:
```
aliases {
	acpi_dev = &some_existing_dev;
};

&some_existing_dev {
	acpi-hid = "PNP...";
};
```

And then in the test case you'd use `DT_ALIAS(acpi_dev)` instead of `DT_NODELABEL(rtc)`

jhedberg replied:
Hmm, I realise one issue with the above is that the DT binding for the device node in question needs to support the `acpi-hid` property, so you'll either need to modify some existing binding to include `acpi-device.yaml` or then you need to modify the binding in this PR.

jhedberg replied:
Btw, since pcie0 already has the acpi-hid property, couldn't you use that? In that case you'd just need to have the `aliases {` section as described above and use `&pci0` for `&some_existing_dev`. This could then later be updated to be some peripheral (once the peripheral DT bindings and drivers have been updated).

jhedberg reviewed (changes requested):
Thanks for the continued work on getting this closer to be merged! There are still a few issues that I see that need sorting out. I would welcome the other reviewers (in particular @dcpleung @tbursztyka and @finikorg) to take another look as well - the PR content has changed quite a lot from its initial iteration.

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -15,6 +15,27 @@
+#define ACPI_RESOURCE_COUNT_GET(res) (res)->mmio_max
+
+enum acpi_res_type {
+	/** IO mapped Resource type */
+	ACPI_RES_TYPE_IO = 1,
```
Ok, I took a look at how you're using these, and it seems the numeric values don't matter. Because of that the `= 1` is misleading, since it gives the impression that it's important that `TYPE_IO` has exactly the value `1`. I'd propose to just remove the `= 1` and let the values be auto-assigned (i.e. starting from `0`).

jhedberg commented on drivers/pcie/host/pcie.c:
```diff
@@ -5,6 +5,14 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#if defined(CONFIG_PCIE_PRT)
+#if defined(CONFIG_X86_64)
+#define DT_DRV_COMPAT intel_pcie
```
It looks like at least arm64 qemu is using `drivers/pcie/host/pcie.c`, however it's using `compatible = "pci-host-ecam-generic"` for its PCIe bus node. This would lead to the conclusion, that `drivers/pcie/host/pcie.c` is not a driver in itself, rather a kind of "generic PCIe code" that can be used by specific PCIe bus drivers? If that's the correct assessment, and we don't want to change this design, then anything that is  arch-specific should be moved into an arch-specific c-file, including all `DT_*()` macro invocations.

tbursztyka replied:
Think is, the dts bindings for intel's pcie has nothing different from pci. In fact, I don't see why intel's pcie controller would have anything specific. 

Anyway, make it  #define DT_DRV_COMPAT pcie_controller

jhedberg replied:
> Anyway, make it  #define DT_DRV_COMPAT pcie_controller

@tbursztyka in that case we also need to update the DTS files to use a different compatible string, and also remove `dts/bindings/pcie/host/intel,pcie.yaml` (which is fine with me - I'm just saying that only changing DT_DRV_COMPAT will not work)

tbursztyka replied:
hum, yes I did not see this comming. But indeed we need pcie,controller.yaml instead of pcie-controller.yaml
and replace all intel,pcie compat strings to pcie,controller.

So far, all about pcie is generic and intel has nothing specific related to it, so we should be good to go in a pcie,controller generic binding (and this would remove this ugly arch specific ifdef that should not be here)

jhedberg replied:
> hum, yes I did not see this coming. But indeed we need pcie,controller.yaml instead of pcie-controller.yaml and replace all intel,pcie compat strings to pcie,controller.

If there's a comma (`,`) present, the first part is expected to be a vendor name, so I'm not sure we can have `pcie` there. Since this is not based on any existing binding, I think we could just use `zephyr` there (like many other Zephyr bindings already do), in which case this would become `zephyr,pcie-controller`. The other alternative would be to leave `pcie-controller.yaml` as-is, since the omission of a vendor and `,` already implies that it's generic.

jhedberg replied:
@tbursztyka @najumon1980 let's keep the `pcie-controller.yaml` name as-is for now, since it's not absolutely necessary to rename it at this point (renaming would also require updating the ECAM binding since it includes the pice-controller one.). So just get rid of the intel binding and extend the pcie-controller one.

jhedberg replied:
You should remove `intel,pcie.yaml`, as suggested. As for ecam, I thought the compat matching was based on the filename and not an explicitly stated compat string inside the yaml file. I'm not sure if it's possible to extend a binding with a compat string with another (i.e. the compatible in a binding overrides the compatible in a binding that it includes). Did you try this? Do you get errors?

jhedberg replied:
..answering myself: it's ok even if the included file has a `compatible: ..` line also. At least I didn't get any errors when I added that to `pice-controller.yaml`.

najumon1980 committed:
tests: remove pci bus compatible id dependency from overlay files

remove pci bus compatible id dependency from qemu platforms overlay
files for disk and ivshmem test apps. This already provided as part of
respecitve board dts files and hence no need to duplicate here.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
docs: rename intel,pcie compatible id to pcie-controller

rename intel,pcie compatible id referance to pcie-controller in
header file comments and rst files

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

jhedberg commented:
@najumon1980 the third test still doesn't run on qemu_x86_64 (it gets skipped). Could you fix that? I would assume that there has to be *some* device available whose HID you can use instead of rtc. I also see that you haven't tried to take advantage of devicetree aliases. You should do that latest at the point when you use different nodes on different platforms, however doing this from the start would be cleaner, since then you can more easily extend the tests in the future.

jhedberg commented:
Regarding this commit (reviewing as a generic comment since GitHub doesn't have a clean way to review commit messages):
> tests: remove pci bus compactible id dependency from overlay files
>
> remove pci bus compactible id dependency from qemu platforms overlay
> files for disk and ivshmem test apps

It's "compatible", not "compactible" (you have the same spelling error in another commit message too, maybe other places as well). Also, it'd be more accurately described as a "property", not "id".

Secondly, please add more information to the commit message, in particular *why* this change is needed. Thanks!

najumon1980 committed:
dts: rtc: mc146818: add acpi hid support in yaml file

add acpi hid support for mc146818 yaml file. Currently this added
for acpi test case support.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

jhedberg commented on drivers/pcie/host/pcie.c:
```diff
@@ -28,6 +30,10 @@ LOG_MODULE_REGISTER(pcie, LOG_LEVEL_ERR);
 #endif
 
+#ifdef CONFIG_PCIE_PRT
+static bool prt_base_irq;
```
You could add a code comment here explaining what this is for (i.e. that ACPI IRQ init was successful).

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -40,6 +61,72 @@ struct acpi_mcfg {
+ * @brief Init legacy interrupt routing table information from acpi.
+ * Currently assume platform have only one PCI bus.
+ *
+ * @param hid the hardware id of the acpi child device
+ * @param uid the unique id of the acpi child device. The uid can be
```
Capitalize ACPI

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -98,10 +175,11 @@ ACPI_RESOURCE *acpi_resource_parse(ACPI_RESOURCE *res, int res_type);
  *
  * @param hid the hardware id of the acpi child device
- * @param inst the unique id of the acpi child device
+ * @param uid the unique id of the acpi child device. The uid can be
+ * NULL if only one device with given hid present in the platform.
```
Capitalize ACPI. I guess it should be HID as well?

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -124,6 +202,24 @@ static inline ACPI_RESOURCE_IRQ *acpi_irq_res_get(ACPI_RESOURCE *res_lst)
+/**
+ * @brief Parse resource table for irq info.
+ *
+ * @param child_dev the device object of the acpi node
```
Capitalize ACPI

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -124,6 +202,24 @@ static inline ACPI_RESOURCE_IRQ *acpi_irq_res_get(ACPI_RESOURCE *res_lst)
+/**
+ * @brief Parse resource table for mmio info.
+ *
+ * @param child_dev the device object of the acpi node
+ * @param mmio_res mmio resource info
```
ACPI, MMIO

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -180,4 +276,15 @@ int acpi_drhd_get(enum AcpiDmarScopeType scope, ACPI_DMAR_DEVICE_SCOPE *dev_scop
 ACPI_MADT_LOCAL_APIC *acpi_local_apic_get(int cpu_num);
+
+/**
+ * @brief invoke an acpi method and return the result.
+ *
+ * @param path the path name of the acpi object
+ * @param arg_list the list of arguments to be pass down
+ * @param ret_obj the acpi result to be return
```
ACPI

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -180,4 +276,15 @@ int acpi_drhd_get(enum AcpiDmarScopeType scope, ACPI_DMAR_DEVICE_SCOPE *dev_scop
+ * @param ret_obj the acpi result to be return
+ * @return return 0 on success or error code
+ */
+int acpi_invoke_method(char *path, ACPI_OBJECT_LIST *arg_list, ACPI_OBJECT *ret_obj);
```
Could it be `const char *path`? Or ACPICA API is missing const so you'd have to typecast internally in that case?

jhedberg commented on include/zephyr/arch/x86/x86_acpi.h:
```diff
@@ -0,0 +1,14 @@
+/**
+ * @brief Encode interrupt flag for x86 architecture.
+ *
+ * @param polarity the interrupt polarity received from acpica lib
+ * @param trigger the interrupt level received from acpica lib
```
ACPICA

jhedberg commented on include/zephyr/dt-bindings/acpi/acpi.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_ACPI_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_ACPI_H_
+
+#define ACPI_IRQ_DETECT 0xFFFFFFFU
+#define ACPI_IRQ_FLAG_DETECT 0xFFFFFFFU
```
I don't see you use these anywhere. I'd suggest to introduce them (and therefore also this file) in a PR where you introduce the first users of them.

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -217,6 +217,9 @@ static ACPI_STATUS dev_resource_enum_callback(ACPI_HANDLE obj_handle, UINT32 lev
 	}
 
 	child_dev = (struct acpi_dev *)&acpi.child_dev[acpi.num_dev++];
+	if (!dev_info->HardwareId.String) {
+		goto exit;
+	}
```
Why is the if-statement *after* the assignment to `child_dev`? Seems like it should be before it, since there is no use of the `child_dev` pointer in the exist code path.

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -345,7 +348,8 @@ int acpi_current_resource_free(ACPI_RESOURCE *res)
 }
 
 #ifdef CONFIG_PCIE_PRT
-static int acpi_get_irq_table(char *bus_name, ACPI_PCI_ROUTING_TABLE *rt_table, uint32_t rt_size)
+static int acpi_get_irq_table(char *bus_name,
+				ACPI_PCI_ROUTING_TABLE *rt_table, size_t rt_size)
```
Please merge this entire function to the end of `acpi_legacy_irq_init()`. It's the only place where you're calling it from and it's already hard-coding the assumption that the `rt_table` parameter is always pointing at `acpi.pci_prt_table` (by iterating of the this table at the end of the function) .

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -455,6 +456,107 @@ ACPI_RESOURCE *acpi_resource_parse(ACPI_RESOURCE *res, int res_type)
+
+	do {
+		if (!res->Length) {
+			LOG_DBG("Error: zero length found!");
```
If this an error (like the log message suggests) why are you using `LOG_DBG()` instead of `LOG_ERR()`?

jhedberg replied:
> There might be device/node with no resource and hence given as only debug info.

Then I'd suggest to remove "Error:" from the message, since that makes it look like an actual error rather than debug info.

jhedberg commented on lib/acpi/acpi.c:
```diff
@@ -455,6 +456,107 @@ ACPI_RESOURCE *acpi_resource_parse(ACPI_RESOURCE *res, int res_type)
+	struct acpi_reg_base *reg_base = mmio_res->reg_base;
+	int mmio_cnt = 0;
+
+	memset(mmio_res, 0, sizeof(*mmio_res));
```
Wouldn't you in principle only need to pre-initialise `mmio_res->mmio_max = 0;`?

jhedberg commented on tests/lib/acpi/boards/qemu_x86_64.overlay:
```diff
@@ -0,0 +1,18 @@
+
+/* TODO: This overlay file need to update once mc146818 driver eneable for acpi support.
+ */
+
```
Having `aliases` here would make sense, something like:
```
aliases {
	acpi_dev = &rtc0;
};
```

jhedberg commented on tests/lib/acpi/boards/qemu_x86_64.overlay:
```diff
@@ -0,0 +1,18 @@
+/* TODO: This overlay file need to update once mc146818 driver eneable for acpi support.
+ */
+
+/ {
+	soc {
+		rtc0: rtc0 {
+			compatible = "motorola,mc146818";
+			acpi-hid = "PNP0B00";
+			status = "okay";
+		};
+	};
+};
```
This looks strange (and odd that it works), since you already have an rtc0 node in the qemu DTS. Instead, can't you just modify the existing node with something like:
```
&rtc0 {
	acpi-hid = "PNP0B00";
};
```

jhedberg commented on tests/lib/acpi/src/main.c:
```diff
@@ -8,6 +8,22 @@
 #include <zephyr/kernel.h>
 #include <zephyr/acpi/acpi.h>
 
+#if (ACPI_DT_HAS_HID(DT_NODELABEL(pcie0)))
```
What you're really interested in is the same condition that the PCIe code itself uses to require a HID property, i.e. `CONFIG_PCIE_PRT`. I think that's what you should check for here.

najumon1980 replied:
This test case nothing to do with PRT. It test acpi_current_resource_get() and PCIe chosen since it will be default available. Updated to use acpi_dev and can be selected based on board configuration/overlay (as you suggested early)

jhedberg commented on tests/lib/acpi/src/main.c:
```diff
@@ -8,6 +8,22 @@
 #include <zephyr/acpi/acpi.h>
 
+#if (ACPI_DT_HAS_HID(DT_NODELABEL(pcie0)))
+#define PCIE_HID ACPI_DT_HID(DT_NODELABEL(pcie0))
+#define PCIE_UID ACPI_DT_UID(DT_NODELABEL(pcie0))
```
Here it would make sense to use `DT_PROP_OR(..., ..., NULL)`, so that you don't need to have the `#else` branch with some arbitrary looking "INVALID" strings (even if you did have them NULL would be better)

jhedberg commented on tests/lib/acpi/src/main.c:
```diff
@@ -8,6 +8,22 @@
+#define PCIE_UID "INVALID"
+#endif
+
+#if (ACPI_DT_HAS_HID(DT_NODELABEL(rtc0)))
+#define RTC_HID ACPI_DT_HID(DT_NODELABEL(rtc0))
+#define RTC_UID ACPI_DT_UID(DT_NODELABEL(rtc0))
```
This should be something like:
```
#if (DT_NODE_EXISTS(DT_ALIAS(acpi_dev)))
#define ACPI_DEV_HID ACPI_DT_HID(DT_ALIAS(acpi_dev))
#define ACPI_DEV_UID ACPI_DT_UID(DT_ALIAS(acpi_dev))
#else
#define ACPI_DEV_HID NULL
#define ACPI_DEV_HID NULL
#endif
```

>

jhedberg commented on tests/lib/acpi/src/main.c:
```diff
@@ -17,14 +33,44 @@ ZTEST(acpi, test_mcfg_table)
+	ACPI_RESOURCE *res_lst;
+	int ret;
+
+	Z_TEST_SKIP_IFNDEF(ACPI_DT_HAS_HID(DT_NODELABEL(pcie0)));
```
The `Z_TEST_SKIP_IFNDEF()` macro is AFAIK only intended to be used for Kconfig options, which are either undefined (if they're set to `n`) or defined (if they're set to `y`). However here you always get something that's defined (either `0` or `1`). I think what this really should be checking for is the same condition that the PCIe code is using to require the ACPI properties or not, i.e. CONFIG_PCIE_PRT:
```c
	Z_TEST_SKIP_IFNDEF(CONFIG_PCIE_PRT);
```

jhedberg replied:
> Nothing to do with PRT. Updated by define APCI_TEST_DEV macro if HID property exist. I observed Z_TEST_SKIP_IFNDEF skip test case if node property does not exist.

Yep. I just checked the implementation, and this is using `COND_CODE_1()` which will only evaluate to "true" if the given macro is both defined *and* has the value `1`. In that sense the `SKIP_IFNDEF()` name is slightly misleading, since it's not just looking for whether the macro is defined or not but also its value.

jhedberg commented on tests/lib/acpi/src/main.c:
```diff
@@ -17,14 +33,44 @@ ZTEST(acpi, test_mcfg_table)
+	struct acpi_mmio_resource mmio_res;
+	int ret;
+
+	Z_TEST_SKIP_IFNDEF(ACPI_DT_HAS_HID(DT_NODELABEL(rtc0)));
```
Are you sure this does the right thing? The `IFNDEF` in the name implies that it's only checking if the given macro is defined or not, however I think the macro will always be defined, either with the value `1` or `0`, so the test would never be skipped.

Instead, you may need to use something like:
```c
	if (!DT_NODE_EXISTS(DT_ALIAS(acpi_dev))) {
		ztest_test_skip();
	}
```

najumon1980 replied:
see my previous comments

najumon1980 commented:
@jhedberg @dcpleung @tbursztyka @finikorg , updated all the review comments, please help to review and merge the PR

najumon1980 commented:
@tbursztyka @dcpleung could you please review and approve if no more change required

tbursztyka commented on drivers/pcie/host/pcie.c:
```diff
@@ -291,8 +298,16 @@ unsigned int pcie_alloc_irq(pcie_bdf_t bdf)
+		}
 #else
 		irq = arch_irq_allocate();
 #endif
```
rewrite the above so it becomes:
```
#ifdef CONFIG_PCIE_PRT
		irq = acpi_legacy_irq_get(bdf);
		if (prt_base_irq) {
			irq = acpi_legacy_irq_get(bdf);
		} else
#else
                {
			irq = arch_irq_allocate();
		}
#endif 
```

jhedberg replied:
> rewrite the above so it becomes:

@tbursztyka I don't think what you proposed will generate correct code, since it leaves out the `arch_irq_allocate()` call completely if `CONFIG_PCIE_PRT` is set. Anyway, think Najumon's original code is actually a bit more readable, even though there`s a `arch_irq_allocate()` in two places in the code.

najumon1980 replied:
> > rewrite the above so it becomes:
> 
> @tbursztyka I don't think what you proposed will generate correct code, since it leaves out the `arch_irq_allocate()` call completely if `CONFIG_PCIE_PRT` is set. Anyway, think Najumon's original code is actually a bit more readable, even though there's `arch_irq_allocate()` in two places.

updated with slight optimization

tbursztyka commented on dts/bindings/pcie/host/intel,pcie.yaml:
```diff
@@ -1,8 +0,0 @@
-# Copyright (c) 2020 Intel Corporation
```
if you remove it, you'll have to affect all the .dtsi/.dts files that use it it this commit and not in the other ones.

tbursztyka commented on boards/x86/qemu_x86/qemu_x86.dts:
```diff
@@ -49,7 +49,8 @@
 		#address-cells = <1>;
 		#size-cells = <1>;
-		compatible = "intel,pcie";
+		compatible = "pcie-controller";
```
so this change should go in the previous patch that remove "intel,pcie"

najumon1980 replied:
Which commit you referring ? This commit is specifically for board configuration changes

jhedberg replied:
> Which commit you referring ? This commit is specifically for board configuration changes

I think @tbursztyka is referring to the commit where you remove the intel,pcie binding (yaml) file. Once that file is removed there shouldn't be any more references to "intel,pcie" in dts files or the documentation.

najumon1980 replied:
> > Which commit you referring ? This commit is specifically for board configuration changes
> 
> I think @tbursztyka is referring to the commit where you remove the intel,pcie binding (yaml) file. Once that file is removed there shouldn't be any more references to "intel,pcie" in dts files or the documentation.

updated

tbursztyka commented on tests/drivers/disk/disk_access/boards/qemu_x86_64.overlay:
```diff
@@ -4,11 +4,6 @@
 
 / {
 	pcie0 {
-		#address-cells = <1>;
```
this patch is fine, but should come in the history **before** the patch that remove "intel,pcie"

najumon1980 replied:
updated

tbursztyka reviewed (changes requested):
Check the phrasing of your commit messages (typos, punctuation, capital letters...), they really need some grooming.

jhedberg commented on lib/acpi/Kconfig:
```diff
@@ -49,3 +43,15 @@ config ACPI_HID_LEN_MAX
 	help
 	  Size of HID string.
+
+config ACPI_IRQ_VECTOR_MAX
```
All ACPI Kconfig symbols should be inside the `if ACPI` block. This includes also the existing HID, however in that case I think you can remove the Kconfig option completely since the ACPI spec defines a maximum worst-case length for this. That said, I also don't see you populating or using the hid value anywhere, so you should probably just remove it until there's code that needs it.

jhedberg commented on west.yml:
There's a typo in the commit message title: upate -> update. Also, the commit prefix for west.yml changes is normally "manifest: ..." instead of "build: .."

jhedberg commented:
@najumon1980 are you planning to continue working on this? CI is currently failing. Would be good to get this cleaned up and merged well before the 3.6 feature freeze (expected to happen in about two weeks)

najumon1980 committed:
tests: lib: acpi: add fake x86 acpi function

add fake x86 acpi function and config macros for acpi unit test

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 commented:
@jhedberg fixed CI jobs which was due to later changes. Could you please help to merge the PR if no more comments

najumon1980 commented on lib/acpi/acpi.c:
```diff
@@ -216,6 +216,10 @@ static ACPI_STATUS dev_resource_enum_callback(ACPI_HANDLE obj_handle, UINT32 lev
 
+	if (!dev_info->HardwareId.String) {
+		goto exit;
+	}
```
updated

najumon1980 commented on include/zephyr/acpi/acpi.h:
```diff
@@ -40,6 +60,72 @@ struct acpi_mcfg {
 } __packed;
 
+struct acpi_irq_resource {
+	uint32_t flags;
+	union {
+		uint16_t irq;
+		uint16_t irqs[CONFIG_ACPI_IRQ_VECTOR_MAX];
+	};
+	uint8_t irq_vector_max;
+};
+
+struct acpi_reg_base {
+	enum acpi_res_type type;
+	union {
+		uintptr_t mmio;
+		uintptr_t port;
+	};
+	uint32_t length;
+};
+
+struct acpi_mmio_resource {
+	struct acpi_reg_base reg_base[CONFIG_ACPI_MMIO_ENTRIES_MAX];
```
agree and yes we can create a separate PR for it.

najumon1980 commented:
@jhedberg  @tbursztyka  @dcpleung  @finikorg could you please approve and help to merge ?

jhedberg reviewed (dismissed):
This seems good enough, assuming we fix the things with the header file that I pointed out in a follow-up PR. It'd be good to get this in before the feature freeze, especially since upstream ACPICA already depends on the new `acpi_timer_get()` API being introduced here.

tbursztyka commented on dts/bindings/acpi/acpi-device.yaml:
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+# Common fields for PCIe devices
```
"Common fields for ACPI informed based devices"

najumon1980 replied:
updated

tbursztyka commented on dts/bindings/acpi/acpi-device.yaml:
```diff
@@ -0,0 +1,23 @@
+# Copyright (c) 2023 Intel Corporation
```
Rename the file to acpi.yaml

najumon1980 replied:
updated

tbursztyka reviewed (changes requested):
still an issue in dts side.
